nexus/src/app/background/tasks/tuf_artifact_replication.rs

```diff
@@ -570,104 +589,168 @@ impl ArtifactReplication {
+            generation,
+            artifacts: inventory.0.keys().map(|h| h.to_string()).collect(),
+        };
+        Ok((config, inventory))
     }
 
-    /// Ask all sled agents to list the artifacts they have, and mark those
-    /// artifacts as present on those sleds.
-    async fn list_artifacts_on_sleds(
-        &mut self,
-        opctx: &OpContext,
-        sleds: &[Sled],
-        inventory: &mut Inventory,
-        counters: &mut TufArtifactReplicationCounters,
-    ) {
-        let responses =
-            futures::future::join_all(sleds.iter().map(|sled| async move {
-                let response = sled.client.artifact_list().await;
-                (sled, Utc::now(), response)
-            }))
-            .await;
-        let mut requests = Vec::new();
-        for (sled, time, response) in responses {
-            let mut error = None;
-            match response {
-                Ok(response) => {
+    /// Put `config` to `sled`, then query `sled` for a list of its artifacts.
+    ///
+    /// Returns [`ControlFlow::Break`] if `config` was rejected by the sled due
+    /// to an invalid generation number, or if the list response contained a
+    /// different generation.
+    async fn sled_put_config_and_list(
+        log: &slog::Logger,
+        sled: &Sled,
+        config: &ArtifactConfig,
+        ringbuf_tx: &mpsc::Sender<TufArtifactReplicationRequest>,
+    ) -> ControlFlow<(), BTreeMap<ArtifactHash, u32>> {
+        let response = sled.client.artifact_config_put(config).await;
+        ringbuf_tx
+            .send(TufArtifactReplicationRequest {
+                time: Utc::now(),
+                target_sled: sled.id,
+                operation: TufArtifactReplicationOperation::PutConfig {
+                    generation: config.generation,
+                },
+                error: response.as_ref().err().map(|err| {
+                    error!(
+                        log,
+                        "Failed to put artifact config";
+                        "error" => InlineErrorChain::new(err),
+                        "sled" => sled.client.baseurl(),
+                        "generation" => &config.generation,
+                    );
+                    err.to_string()
+                }),
+            })
+            .await
+            .ok();
+        // Bail without sending a list request if the config put failed.
+        if let Err(err) = response {
+            // If the request failed because the sled told us the
+            // generation was invalid, return `Break`.
+            if let sled_agent_client::Error::ErrorResponse(response) = err {
+                if response.status() == StatusCode::CONFLICT
+                    && response.error_code.as_deref()
+                        == Some("CONFIG_GENERATION")
+                {
+                    return ControlFlow::Break(());
+                }
+            }
+            return ControlFlow::Continue(BTreeMap::new());
+        }
+
```

Comment:
It's possible that the sled-agent crashes here. If we don't persist the artifact list with the generation, then I believe we'll get back an `Error::NoConfig` upon restart, if everything on the sled-agent side comes back in time. In this case we'll wait for the next call to the background task. This seems safe, but also a bit unnecessary if this race did occur. Instead the ledger could be read and the artifact list loaded directly from the file instead, if the ledger was always written *before* the watch config was updated.